app/src/main/java/net/gsantner/markor/activity/NoteActivity.java

```diff
@@ -206,47 +207,39 @@ protected void onPause() {
 
     private void setupKeyboardBar() {
         if (_appSettings.isShowMarkdownShortcuts() && _markdownCharBar.getChildCount() == 0) {
-            appendRegularShortcuts();
-            if (_appSettings.isSmartShortcutsEnabled()) {
-                appendSmartBracketShortcuts();
-            } else {
-                appendRegularBracketShortcuts();
-            }
-            for (String shortcut : Constants.KEYBOARD_SHORTCUTS_MORE) {
-                appendButton(shortcut, new KeyboardBarListener());
-            }
+            appendSmartActions();
+            appendRegularActions();
+            appendExtraActions();
         } else if (!_appSettings.isShowMarkdownShortcuts()) {
             findViewById(R.id.note__activity__scroll_markdownchar_bar).setVisibility(View.GONE);
         }
     }
 
-    private void appendRegularShortcuts() {
-        for (String shortcut : Constants.KEYBOARD_SHORTCUTS) {
-            appendButton(shortcut, new KeyboardBarListener());
+    private void appendRegularActions() {
+        for (int[] actions : Constants.KEYBOARD_REGULAR_ACTIONS_ICONS) {
+            appendButton(actions[0], new KeyboardRegularActionListener(Constants.KEYBOARD_REGULAR_ACTIONS[actions[1]]));
         }
     }
 
-    private void appendRegularBracketShortcuts() {
-        for (String shortcut : Constants.KEYBOARD_SHORTCUTS_BRACKETS) {
-            appendButton(shortcut, new KeyboardBarListener());
+    private void appendSmartActions() {
+        for (int[] actions : Constants.KEYBOARD_SMART_ACTIONS_ICON) {
+            appendButton(actions[0], new KeyboardSmartActionsListener(Constants.KEYBOARD_SMART_ACTIONS[actions[1]]));
         }
-
     }
 
-    private void appendSmartBracketShortcuts() {
-        for (String shortcut : Constants.KEYBOARD_SMART_SHORTCUTS) {
-            appendButton(shortcut, new KeyboardBarSmartShortCutListener());
-        }
+    private void appendExtraActions(){
+        for (int[] actions : Constants.KEYBOARD_EXTRA_ACTIONS_ICONS) {
+                appendButton(actions[0], new KeyboardRegularActionListener(Constants.KEYBOARD_EXTRA_ACTIONS[actions[1]]));
```

Comment:
doesnt this append always the same?

Reply:
yeah , i made a third category for actions which doesn't have a consistent syntax , such as `()[]` and `----` . Thought if in the future , actions for tables or images were brought , a third cateogry would be apt for such syntax's . Right now it uses the same keyboard listener of the Regular Action.

Reply:
Yeah, even if they use the same, wouldn't it be better to append the current iterated action to ensure not making bugs later?
e.g.

```
for (int i=0; i < actions.length ; i+=2) {
append(action[i] , .... action[i+1];
}
```

Reply:
I didnt understand . All the actions are appended exactly once to the keyboard bar , isn't that the desired outcome ? . The constants KEYBOARD_EXTRA_ACTIONS_ICONS is of type `int[][]` , used to store a drawable and an int (which referes to its respective posiiton of character to be appended).

Reply:
I meant that you are iterating over actions[] but yoi always Insert actions[0] and action[1]

Reply:
Oh, its a multi array, right, my bad :D

Reply:
Yeah :-)